File: codesignal/combinationSum.js
// first pass solution
function combinationSum(a, sum) {
    let set = new Set();
    a.sort((a, b) => a - b);
    
    function DFS(a, sum, currSum, path, index) {
        if (currSum === sum) {
            set.add('(' + path.join(' ') + ')');
            return;
        }
        
        for (let i = index; i < a.length; i++) {
            if (currSum + a[i] <= sum) {
                DFS(a, sum, currSum + a[i], [...path, a[i]], i);
            }
        }
    }
    
    DFS(a, sum, 0, [], 0);
    
    return set.size > 0 ? Array.from(set).join("") : 'Empty';
}

File: codesignal/combinationSum.java
String combinationSum(int[] a, int sum) {
    StringBuilder sb = new StringBuilder();
    Set<String> set = new LinkedHashSet<>();
    
    Arrays.sort(a);
    
    DFS(a, sum, 0, set, "");
    
    for (String str : set) {
        sb.append(str);
    }
    
    return sb.length() == 0 ? "Empty" : sb.toString();
}

void DFS(int[] a, int sum, int index, Set<String> set, String path) {
    if (sum == 0) {
        set.add("(" + path.trim() + ")");
        return;
    }
    
    for (int i = index; i < a.length; i++) {
        if (sum - a[i] >= 0) {
            DFS(a, sum - a[i], i, set, path + " " + a[i]);
        }
    }
}
